### ISSUE TITLE
Trailing commas in select statement should raise linting error

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#362]
#354 allows parsing of trailing commas in select statements. This is a bigquery feature only (and so technically should raise a *parsing* error), but on balance I think it's more _useful_ that the query parses and then this can be raised as an issue.

I think even in dialects where this is supported, this should be regarded as an anti-pattern. We should have a linting rule which catches this.

**Comment by aaronsteers** [comment]
@alanmcruickshank - I'm new to the science and practice behind linting but this looks like a good first issue and I'd love the opportunity to contribute if you think this would be a good one to hand off.

If so - and if it would be helpful to start up a PR on this one - do you have any guidance on specific rules that might be similar-enough to this one that I could adapt without having to build completely from scratch?

**Comment by pwildenhain** [comment]
@aaronsteers as someone who has recently contributed, I would say the most helpful thing to wrap my head around first, was how the parsing breaks up the sql. For example

test.sql
```sql
select
    col1,
    col2
from my_table
```

parse it
```sh
$ sqlfluff parse test.sql
=== [test.sql] ===
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                newline_RawSegment:                         [6](1, 1, 7)        '\n'
                whitespace_RawSegment:                      [7](1, 2, 1)        '    '
                SelectTargetElementSegment:                 [11](1, 2, 5)
                    ObjectReferenceSegment:                 [11](1, 2, 5)
                        naked_identifier_ReSegment:         [11](1, 2, 5)       'col1'
                comma_KeywordSegment:                       [15](1, 2, 9)       ','
                newline_RawSegment:                         [16](1, 2, 10)      '\n'
                whitespace_RawSegment:                      [17](1, 3, 1)       '    '
                SelectTargetElementSegment:                 [21](1, 3, 5)
                    ObjectReferenceSegment:                 [21](1, 3, 5)
                        naked_identifier_ReSegment:         [21](1, 3, 5)       'col2'
                [META] Dedent:                              [25](1, 3, 9)
            newline_RawSegment:                             [25](1, 3, 9)       '\n'
            FromClauseSegment:                              [26](1, 4, 1)
                FROM_KeywordSegment:                        [26](1, 4, 1)       'from'
                [META] Indent:                              [30](1, 4, 5)
                whitespace_RawSegment:                      [30](1, 4, 5)       ' '
                TableExpressionSegment:                     [31](1, 4, 6)
                    ObjectReferenceSegment:                 [31](1, 4, 6)
                        naked_identifier_ReSegment:         [31](1, 4, 6)       'my_table'
                [META] Dedent:                              [39](1, 4, 14)
    newline_RawSegment:                                     [39](1, 4, 14)      '\n'
```

I would approach this by determining if there is a `comma_KeywordSegment` _after_ the last `SelectTargetElementSegment`, _within_ the `SelectClauseSegment`.

A helpful rule to look at might be [`Rule_L013`](https://github.com/sqlfluff/sqlfluff/blob/7d68d018cc84ef7aaf3cdd777365a306cbeb6053/src/sqlfluff/rules/std.py#L1266) ([docs](https://docs.sqlfluff.com/en/latest/rules.html#sqlfluff.rules.std.Rule_L013)) since it checks to see if columns are explicitly aliased, which is sort of similar to this.

**Comment by NiallRees** [comment]
I'm actually not convinced that this should be a linting error. I wish all dialects supported this (Python allows this in lists!). But the benefit of configurable rules is that I don't have to be 😄

**Comment by pwildenhain** [comment]
I can see your point, I do like that Python allows this. One potential argument against this is cross database compatibility, but BigQuery seems so different anyway, that I'm not sure that argument holds water.

Anyone else want to chime in on why this should be a linting error?

**Comment by aaronsteers** [comment]
On platforms where this would fall compilation, the linter could be a help to prevent/detect later failure. In my opinion, this is less of a style guide issue and more of straight-up syntax check - which is just as valid, I think, for linters in general.

I'd agree that for platforms (dialects) which support it, definitely the best practice will be to keep the trailing comma. (Except in cases where portability to another platform is higher priority than overall style.) For good reason, many json parsers, and the python language itself have been evolving to accept (and prefer) a trailing comma in lists. It makes code reviews cleaner and significantly reduces diffs on lines that don't actually have substantive changes on them.

That said, until this is broadly adopted by SQL platforms, I do think think it would be valuable to detect and resolve, at dev-time, syntax errors that otherwise would become errors at runtime.

Just my two cents. 😅

**Comment by dhimmel** [comment]
> for platforms (dialects) which support it, definitely the best practice will be to keep the trailing comma. (Except in cases where portability to another platform is higher priority than overall style.)

Well said. At @related-sciences, we primarily use GoogleSQL on BigQuery, so the trailing commas are much appreciated.

If anyone finds this issue looking how to require the select clause trailing comma via a `pyproject.toml` configuration file:

```toml
[tool.sqlfluff.rules.convention.select_trailing_comma]
select_clause_trailing_comma = "require"  # CV03 / L038
```
